aeppl/truncation.py

```diff
@@ -0,0 +1,129 @@
+import warnings
+from typing import List, Optional
+
+import aesara.tensor as at
+import numpy as np
+from aesara.assert_op import Assert
+from aesara.graph.basic import Node
+from aesara.graph.fg import FunctionGraph
+from aesara.graph.opt import local_optimizer
+from aesara.scalar.basic import Clip
+from aesara.tensor.elemwise import Elemwise
+from aesara.tensor.random.op import RandomVariable
+from aesara.tensor.var import TensorConstant
+
+from aeppl.logprob import _logcdf, _logprob
+
+
+# TODO: Add interval transform
+class CensoredRV(RandomVariable):
+    r"""A base class for censored `RandomVariable`\s."""
+
+    def __init__(self, base_op):
+        self.base_op = base_op
+        super().__init__(
+            "censored",
+            base_op.ndim_supp,
+            list(base_op.ndims_params) + [base_op.ndim_supp] * 2,
+            base_op.dtype,
+            inplace=False,
+        )
+
+
+@_logprob.register(CensoredRV)
+def censor_logprob(op, value, *inputs, name=None, **kwargs):
+
+    *rv_params, lower_bound, upper_bound = inputs
+    logprob = _logprob(op.base_op, value, *rv_params, name=name, **kwargs)
+    logcdf = _logcdf(op.base_op, value, *rv_params, name=name, **kwargs)
+    if op.base_op.name:
+        logprob.name = f"{op.base_op.name}_logprob"
+        logcdf.name = f"{op.base_op.name}_logcdf"
+
+    is_lower_bounded, is_upper_bounded = False, False
+    if not (
+        isinstance(upper_bound, TensorConstant) and np.all(upper_bound.value == np.inf)
+    ):
+        is_upper_bounded = True
+
+        logccdf = at.log(1 - at.exp(logcdf))
+        # For right censored discrete RVs, we need to add an extra term
+        # corresponding to the pmf at the upper bound
+        if op.base_op.dtype == "int64":
+            # TODO: Replace by at.logaddexp in next release of Aesara
+            logccdf = at.log(at.exp(logccdf) + at.exp(logprob))
+
+        logprob = at.switch(
+            at.eq(value, upper_bound),
+            logccdf,
+            at.switch(at.gt(value, upper_bound), -np.inf, logprob),
+        )
+    if not (
+        isinstance(lower_bound, TensorConstant) and np.all(lower_bound.value == -np.inf)
+    ):
+        is_lower_bounded = True
+        logprob = at.switch(
+            at.eq(value, lower_bound),
+            logcdf,
+            at.switch(at.lt(value, lower_bound), -np.inf, logprob),
+        )
+
+    if is_lower_bounded and is_upper_bounded:
+        logprob = Assert("lower_bound <= upper_bound")(
+            logprob, at.all(at.le(lower_bound, upper_bound))
+        )
+
+    return logprob
+
+
+@local_optimizer(tracks=[Elemwise])
+def censor_rvs(fgraph: FunctionGraph, node: Node) -> Optional[List[Node]]:
+    # TODO: Canonicalize x[x>ub] = ub -> clip(x, x, ub)
+
+    rv_map_feature = getattr(fgraph, "preserve_rv_mappings", None)
```

Comment:
Can we just let this kind of check fail instead of using the `getattr`?

Reply:
No, we need to return `None` to signify that the optimization is skipped.  I think the optimization framework will allow uncaught exceptions, but they'll show up as warnings, and we don't want that.  

Plus, I believe there are still a few places where exceptions are used for control flow during optimization, and we still need to fix that.